src/Spectre.Data/Datasets/BasicTextDataset.cs

```diff
@@ -0,0 +1,307 @@
+                        if (!int.TryParse((metadata.Length>2)?metadata[2]:null, NumberStyles.Any, CultureInfo.InvariantCulture, out z))
+                            z = -1;
+
+                        _spatialCoordinates.Add(new SpatialCoordinates(x, y, z));
+                        _intensity.Add(new double[_mz.Length]);
+
+                        int backIdx = _intensity.Count - 1;
+                        for (int i = 0; i < intensities.Length; i++)
+                            if (!double.TryParse(intensities[i], NumberStyles.Any, CultureInfo.InvariantCulture, out _intensity[backIdx][i]))
+                                _intensity[backIdx][i] = double.NaN;
+                            
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                throw new InvalidDataException("Error while parsing " + filePath + " file.", e);
+            }
+        }
+        /// <summary>
+        /// See <see cref="IDataset"/> for description.
+        /// </summary>
+        /// <exception cref="InvalidDataException">Throws when the data is null or 
+        /// the length of data is not matching the dataset length.</exception>
+        public void AppendFromRawData(double[,] data)
+        {
+            if(data == null)
+                throw new InvalidDataException("The input data is null.");
+            if(_mz.Length != data.GetLength(1))
+                throw new InvalidDataException("The length of input data does not match the length of present data.");
+            for (int i = 0; i < data.GetLength(0); i++)
+            {
+                _intensity.Add(new double[data.GetLength(1)]);
+                _spatialCoordinates.Add(new SpatialCoordinates(-1, -1, -1));
+                int backIdx = _intensity.Count - 1;
+                for (int j = 0; j < data.GetLength(1); j++)
+                    _intensity[backIdx][j] = data[i, j];
+            }
+        }
+        /// <summary>
+        /// See <see cref="IDataset"/> for description.
+        /// </summary>
+        public DataPoint GetDataPoint(int spectrumIdx, int valueIdx)
+        {
+            return new DataPoint(_mz[valueIdx], _intensity[spectrumIdx][valueIdx]);
+        }
+        /// <summary>
+        /// See <see cref="IDataset"/> for description.
+        /// </summary>
+        public DataPoint[] GetDataPoints(int spectrumIdx, int valueIdxFrom, int valueIdxTo)
+        {
+            if (valueIdxFrom >= valueIdxTo)
+                throw new IndexOutOfRangeException();
+            int valueCnt = valueIdxTo - valueIdxFrom;
+
+            DataPoint[] result = new DataPoint[valueCnt];
+            for (int i = 0; i < valueCnt; i++)
+                result[i] = new DataPoint(_mz[valueIdxFrom + i], _intensity[spectrumIdx][valueIdxFrom + i]);
+            return result;
+        }
+
+        /// <summary>
+        /// See <see cref="IDataset"/> for description.
+        /// </summary>
+        public int GetSpectrumLength()
```

Comment:
This remains constant since initialization. Maybe use property without `set` to assign it only in the constructor and leave constant?